src/toxcore/net_crypto/net_crypto.rs

```diff
@@ -0,0 +1,177 @@
+        let connections = self.connections.read();
+        let keys_by_addr = self.keys_by_addr.read();
+        let connection = keys_by_addr.get(&(addr.ip(), addr.port())).cloned().and_then(|pk|
+            connections.get(&pk)
+        );
+        match connection {
+            Some(connection) => Box::new(future::result(
+                connection.write().handle_cookie_response(packet, &self.symmetric_key)
+            )),
+            None => Box::new(future::err(
+                Error::new(
+                    ErrorKind::Other,
+                    format!("No crypto connection for address {}", addr)
+                )
+            ))
+        }
+    }
+
+    /// Handle `CryptoHandshake` packet received from UDP socket
+    pub fn handle_udp_crypto_handshake(&self, packet: CryptoHandshake, addr: SocketAddr) -> IoFuture<()> {
+        let connections = self.connections.read();
+        let keys_by_addr = self.keys_by_addr.read();
+        let connection = keys_by_addr.get(&(addr.ip(), addr.port())).cloned().and_then(|pk|
+            connections.get(&pk)
+        );
+        match connection {
+            Some(connection) => Box::new(future::result(
+                connection.write().handle_crypto_handshake(packet, &self.symmetric_key)
+            )),
+            None => Box::new(future::err( // TODO: create crypto connection
+                Error::new(
+                    ErrorKind::Other,
+                    format!("No crypto connection for address {}", addr)
+                )
+            ))
+        }
+    }
+
+    /// Send packet to crypto connection choosing TCP or UDP protocol
+    fn send_packet(&self, packet: DhtPacket, connection: &CryptoConnection) -> IoFuture<()> {
+        //TODO: send to TCP as well
+        if let Some(saddr) = connection.udp_saddr {
+            (self.send_to_udp)(packet, saddr)
+        } else {
+            Box::new(future::err(
+                Error::new(ErrorKind::Other, "Can't send net crypto packet")
+            ))
+        }
+    }
+
+    /// Send `CookieRequest` or `CryptoHandshake` packet if needed depending on
+    /// connection status and update sent counter
+    fn send_status_packet(&self, connection: &mut CryptoConnection) -> IoFuture<()> {
+        match connection.packet_to_send() {
+            Some(packet) => self.send_packet(packet, connection),
+            None => Box::new(future::ok(())),
+        }
+    }
+
+    /// Remove timed out connections
+    pub fn remove_timed_out(&self) {
+        self.connections.write().retain(|_, connection| !connection.read().is_timed_out());
+    }
+
+    /// The main loop that should be run at least 20 times per second
+    pub fn main_loop(&self) -> IoFuture<()> {
+        let connections = self.connections.read();
+        let mut send_futures = Vec::new();
+        // Only one cycle over all connections to prevent many lock acquirements
+        for connection in connections.values() {
```

Comment:
connections.values().iter map collect?

Reply:
No so easy. If change this code to `map` with closure this closure will require `&self` argument to live longer then method body:
```
error[E0495]: cannot infer an appropriate lifetime for lifetime parameter in function call due to conflicting requirements
```